### ISSUE TITLE
Entities using ChangeTrackingStrategy.ChangedNotifications throw an exception when context.Entry(entity) is called

### ISSUE THREAD

**Issue by JonPSmith** [dotnet/efcore#21835]
I'm writing about `INotifyPropertyChanged` on entities that are configured as `.HasChangeTrackingStrategy(ChangeTrackingStrategy.ChangedNotifications)`  and my unit tests throw an exception when I call `context.Entry(entity)`. 

Looking at the exception it looks like `context.Entry(entity)` is calling `DetectChanges` and failing there.

Also, this exception is in EF Core 3.1, but doesn't in EF Core 2.1

### Steps to reproduce

```c#
class Program
{
    static void Main(string[] args)
    {
        TestNotifyModifiedOk();
    }

    public class NotificationEntity : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler PropertyChanged;

        protected void SetWithNotify<T>(T value, ref T field,
            [CallerMemberName] string propertyName = "") 
        {
            if (!Equals(field, value)) 
            {
                field = value; 
                PropertyChanged?.Invoke(this, 
                    new PropertyChangedEventArgs(propertyName)); 
            }
        }
    }

    public class NotifyOne
    {
        public int Id { get; set; }

        public int MyInt { get; set; }

        public int NotifyEntityId { get; set; }
    }

    public class NotifyMany
    {
        public int Id { get; set; }

        public int NotifyEntityId { get; set; }
    }

    public class NotifyEntity : NotificationEntity
    {
        private int _id; 
        private string _myString; 
        private NotifyOne _oneToOne; 

        public int Id
        {
            get => _id;
            set => SetWithNotify(value, ref _id); 
        }

        public string MyString
        {
            get => _myString;
            set => SetWithNotify(value, ref _myString); 
        }

        public NotifyOne OneToOne
        {
            get => _oneToOne;
            set => SetWithNotify(value, ref _oneToOne); 
        }

        public ObservableCollection<NotifyMany> Many { get; } 
            = new ObservableCollection<NotifyMany>(); 
    }

    public class Chapter11DbContext : DbContext 
    {

        public Chapter11DbContext(DbContextOptions<Chapter11DbContext> options) 
            : base(options)
        {
        }

        public DbSet<NotifyEntity> Notify { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<NotifyEntity>()
                .HasChangeTrackingStrategy(ChangeTrackingStrategy.ChangedNotifications);
        }
    }

    public static void TestNotifyModifiedOk()
    {
        //SETUP
        var options = SqliteInMemory.CreateOptions<Chapter11DbContext>();
        using (var context = new Chapter11DbContext(options))
        {
            context.Database.EnsureCreated();

            var entity = new NotifyEntity {MyString = "Test"};
            context.Add(entity);
            context.SaveChanges();
        }

        using (var context = new Chapter11DbContext(options))
        {
            //ATTEMPT
            var entity = context.Notify.First();
            entity.MyString = "Changed";
            context.SaveChanges();
        }

        using (var context = new Chapter11DbContext(options))
        {
            //VERIFY
            var entity = context.Notify.First();
            entity.MyString.ShouldEqual("Changed");
            //see https://github.com/dotnet/efcore/issues/21835
            var xx = context.Entry(entity); //THROWS EXCEPTION HERE
            context.Entry(entity).Property(nameof(NotifyEntity.MyString)).OriginalValue.ShouldEqual("Changed");
        }
    }
}

.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="EfCore.TestSupport" Version="3.2.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="3.1.7" />
  </ItemGroup>

</Project>
```
```
Exception is
```
System.IndexOutOfRangeException
  HResult=0x80131508
  Message=Index was outside the bounds of the array.
  Source=Microsoft.EntityFrameworkCore
  StackTrace:
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.Snapshot`2.get_Item(Int32 index)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.RelationshipsSnapshot.GetValue(InternalEntityEntry entry, IPropertyBase propertyBase)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.GetRelationshipSnapshotValue(IPropertyBase propertyBase)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectNavigationChange(InternalEntityEntry entry, INavigation navigation)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.LocalDetectChanges(InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.EntityEntry.DetectChanges()
   at Microsoft.EntityFrameworkCore.DbContext.TryDetectChanges(EntityEntry entry)
   at Microsoft.EntityFrameworkCore.DbContext.Entry[TEntity](TEntity entity)
   at TestEfCoreIssue21835.Program.TestNotifyModifiedOk() in C:\Users\jonsm\Source\Repos\TestEfCoreIssue21835\Program.cs:line 124
   at TestEfCoreIssue21835.Program.Main(String[] args) in C:\Users\jonsm\Source\Repos\TestEfCoreIssue21835\Program.cs:line 16
```

- The same code that failed works with `ChangeTrackingStrategy.ChangingAndChangedNotifications` config
- This exception is in EF Core 3.1, but doesn't in EF Core 2.2

### Further technical details

EF Core version:
Database provider: Microsoft.EntityFrameworkCore.Sqlite
Target framework: EF Core 3.1.7, and EF Core 5, preview 7
Operating system: Windows 10
IDE: Visual Studio 2019 16

**Comment by ajcvickers** [comment]
@JonPSmith This looks like a bug. However, change tracking proxies are intended to use both changing and changed notifications. It's likely that we will just block using other options.

**Comment by JonPSmith** [comment]
Well, I'm glad it wasn't me reporting something that was my fault. Been doing that lately (sorry about that).

OK. Are you saying that `ChangeTrackingStrategy.ChangedNotifications` shouldn't be used? From my point of view the `ChangeTrackingStrategy.ChangedNotifications` is a) easier to code and b) works with transactions. 

Happy to edit my book to remove the `ChangeTrackingStrategy.ChangedNotifications` version but I suspect its the main way that people handle the problem of `DetectChanges` is too slow. The extra work for `ChangeTrackingStrategy.ChangingAndChangedNotifications` only saves memory.

Also the new Proxy approach to ChangeTrackingStrategy is much easier, but you have to change every entity class. I personally would add `ChangeTrackingStrategy.ChangedNotifications` to just the entities that are a problem if I had to (haven't done that in a client's app, but that's because I try to make sure I don't get into that situation 😄).

**Comment by JonPSmith** [comment]
@ajcvickers 

BTW. The original code came from the 1st edition of my book. I tried it and
- EF Core 2.1 was fine
- EF Core 3.0 fails with the same error

If no one has reported it, then maybe its not important. I just wanted to check on the snapshot. For now I will comment out those tests in the unit test.

**Comment by ajcvickers** [comment]
@JonPSmith Sorry, I misread this--I thought you were talking about proxies. I'll take another look.

**Comment by ajcvickers** [comment]
Marking this as poachable and removing from milestone. This may be puntable--I haven't investigated very deeply.

**Comment by AndriySvyryd** [comment]
@JonPSmith Can you review the repro code? I don't get the exception even on preview7

**Comment by JonPSmith** [comment]
Hi @AndriySvyryd,

To double check the code I created a competently separate console app to run the test. This showed that I have missed part of the entity class in my original submission (the `public ObservableCollection<NotifyMany> Many { get; } ` property), and it looks like this causes the problem. 

I have updated the original comment with this console app, so it should fail now.

*PS. Sorry for the delay in getting back. I wrote a comment last week but must have forgotten to click the Comment button!*
